lib/rubygems/ext/cmake_builder.rb

```diff
@@ -1,21 +1,114 @@
+#  --preset <preset_name> to use a preset
+#
+# If the Gem author provides presets, via CMakePresets.json file, you will likely want to us use one of them.
+# If not, then you may wish to specify a generator (Ninja is recommended since it can build projects in parallel
+# versus a Make generator which build ins serial and thus is *much* slower).
+
+require "fileutils"
+
 class Gem::Ext::CmakeBuilder < Gem::Ext::Builder
-  def self.build(extension, dest_path, results, args=[], lib_dir=nil, cmake_dir=Dir.pwd,
-    target_rbconfig=Gem.target_rbconfig)
+  attr_accessor :runner, :profile
+
+  def initialize
+    @runner = self.class.method(:run)
+    @profile = :release
+  end
+
+  def build(extension, gem_dir, results, args = [], lib_dir = nil, cmake_dir = Dir.pwd, target_rbconfig = Gem.target_rbconfig)
     if target_rbconfig.path
       warn "--target-rbconfig is not yet supported for CMake extensions. Ignoring"
     end
 
-    unless File.exist?(File.join(cmake_dir, "Makefile"))
-      require_relative "../command"
-      cmd = ["cmake", ".", "-DCMAKE_INSTALL_PREFIX=#{dest_path}", *Gem::Command.build_args]
+    # Make sure lib dir is set
+    lib_dir ||= File.join(gem_dir, "lib")
 
-      run cmd, results, class_name, cmake_dir
-    end
+    # Figure the build dir
+    build_dir = File.join(cmake_dir, "build")
 
-    make dest_path, results, cmake_dir, target_rbconfig: target_rbconfig
+    # Check if the gem defined presets
+    check_presets(cmake_dir, args)
+
+    # Configure
+    configure(cmake_dir, build_dir, lib_dir, args, results)
+
+    # Compile
+    compile(cmake_dir, build_dir, args, results)
 
     results
   end
+
+  def configure(cmake_dir, build_dir, lib_dir, args, results)
+    cmd = ["cmake",
+           cmake_dir,
+           "-B",
+           build_dir,
+           "-DCMAKE_RUNTIME_OUTPUT_DIRECTORY=#{lib_dir}", # Windows
+           "-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=#{lib_dir}", # Not Windows
+           *Gem::Command.build_args,
+           *args]
+
+    runner.call(cmd, results, "cmake", cmake_dir)
+  end
+
+  def compile(cmake_dir, build_dir, args, results)
+    cmd = ["cmake",
+           "--build",
+           build_dir.to_s,
+           "--config",
+           @profile.to_s]
+    runner.call(cmd, results, "cmake", cmake_dir)
+  end
+
+  private
+
+  def check_presets(cmake_dir, args)
+    # Return if the user specified a preset
+    return unless args.grep(/--preset/i).empty?
+
+    cmd = ["cmake",
+           "--list-presets"]
+
+    begin
+      require "open3"
+      build_env = { "SOURCE_DATE_EPOCH" => Gem.source_date_epoch_string }
+      stdout, _, _ = Open3.capture3(build_env, *cmd, chdir: cmake_dir)
+      message = <<~EOS
+        The gem author provided a list of presets that can be used to build the gem. To use a preset specify it on the command line:
+
+          gem install <gem_name> -- --preset <preset_name>
+
+        #{stdout}
+      EOS
+
+      STDOUT << message << "\n"
```

Comment:
What's the reason for executing `cmake` manually here rather than delegating to the `runner`? This kind of output would be normally appended to `results` (so it's later displayed on errors or when verbose mode is enabled) rather than printed to `STDOUT` directly. Also, because of not delegating to the runner, I believe the `rescue Gem::InstallError` can never rescue anything.

Reply:
Ok, updated to use `runner` and not to use `STDOUT`.